common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -405,19 +469,35 @@ protected boolean runAllTasks(long timeoutNanos) {
             if ((runTasks & 0x3F) == 0) {
                 lastExecutionTime = ScheduledFutureTask.nanoTime();
                 if (lastExecutionTime >= deadline) {
+                    result = DeadlineRunResult.TIMED_OUT;
```

Comment:
> the diff makes it a little difficult to read but this is inside runAllTasksFrom() method which does not call afterRunningAllTasks

Understood. It seems like there is duplication in the return status of this method and the `executionTimeCallback` which can also be used to calculate timeout. What I'm wondering is can we just change the return type of this method to `boolean` and rely upon `executionTimeCallback` to determine if a timeout occurred? We should be able to get rid of `afterRunningAllTasks` also.

```java
// SingleThreadEventLoop.java

@Override
protected boolean runAllTasks() {
  super.runAllTasks();
  // also run tail tasks
}

@Override
protected boolean runAllTasks(long timeoutNanos) {
  // do what we do now...

  // instead of return super.runAllTasks(runTimeout); do
  super.runAllTasks(runTimeout, runTimeoutUpdater);

  // runTimeout now indicates if a timeout occurred
  // no need to rely on the base class calling afterRunningAllTasks
  lastRunTimedOut = runTimeout <= 0;

  if (!lastRunTimedOut) {
    runAllTasksFrom(tailTasks);
     // we arguably should use the remaining runTimeout here ... and then
     // set lastRunTimedOut if necessary after this call.
  }
}

// SingleThreadEventExecutor.java

protected boolean runAllTasks(long timeoutNanos, LongBiConsumer executionTimeCallback) {
  return runAllTasksFrom(taskQueue, timeoutNanos, executionTimeCallback);
}

protected boolean runAllTasksFrom(Queue<Runnable> taskQueue, long timeoutNanos,
                                                LongBiConsumer executionTimeCallback) {
  // return true if any tasks were run
}
```